### PR TITLE
WIP glob-19200 make logging a factory

### PR DESCRIPTION
**Pull request by cdm-ium**
Still a WIP but I want to show what's going on as I hvae this working in leif and am ready for some feedback.

I want to:
- add more tests that involve instantiating a graph and a logger attached to the graph.
- fix small details e.g. automatically naming the tag sent to loggly correctly.
- step back and see if the module structure here makes sense/think about how to have it make more sense now that we brought a lot of code out of the gateway into here.

### CODE REVIEW

File: dist/index.js [hunk 1]
```diff
@@ -1,19 +0,0 @@
-'use strict';
```
Comment (cdm-ium) [inline dist/index.js]: I removed the `dist` package altogether. Our other node projects didn't have it checked in...

File: src/defaults.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        '/gql/health',
+    ],
+    console: {
+        colorize: false,
+    },
+    level: 'info',
+    loggly: {
+        enabled: false,
+        environment: 'dev',
+        tagName: 'leif', // xxx set tag name properly
```
Comment (cdm-ium) [inline src/defaults.js]: I need to fix this. Should be easy enough to set this properly from the logger factory.

File: src/logger.js [hunk 3]
```diff
@@ -45,4 +69,105 @@ export default once((config) => {
...
+            return logger.log(config.level, logEntry.message, logEntry);
+        },
+    };
+}
+
+
+@binding('logger')
+@defaults(loggingDefaults)
+class Logger {
+    constructor(graph) {
```
Comment (cdm-ium) [inline src/logger.js]: This is the place to start. All our logging functionality is invoked from here.

File: src/logger.js [hunk 4]
```diff
@@ -45,4 +69,105 @@ export default once((config) => {
...
+                ? omit(req.headers, config.omitReqProperties) : {};
+            return JSON.stringify(headers);
+        });
+
+        const format = json(config.morgan.format);
+        const options = {
+            stream: addStream(this.baseLogger, config),
+            skip: skip(config),
+        };
+        graph.app.use(morgan(format, options));
```
Comment (cdm-ium) [inline src/logger.js]: I don't love that I attached the `app` to the graph in my main `index.js` but it lets you add midlewares easily enough....

File: src/middleware.js [hunk 5]
```diff
@@ -1,54 +0,0 @@
-import morgan from 'morgan';
```
Comment (cdm-ium) [inline src/middleware.js]: There is no more middleware that `nodule-logging` exports. Instead the `logger` factory sets up a logging midleware on the app for you.

File: src/logger.js [hunk 6]
```diff
@@ -45,4 +69,107 @@ export default once((config) => {
...
+        morgan.token('message', req => req.name || '-');
+        morgan.token('request-id', req => req.id);
+        morgan.token('request-headers', (req) => {
+            const headers = this.config.includeReqHeaders === true
+                ? omit(req.headers, this.config.omitReqProperties) : {};
+            return JSON.stringify(headers);
+        });
+    }
+
+    makeMiddleware() {
```
Comment (cdm-ium) [inline src/logger.js]: I removed the dependency on the app from this component. This function is something you'd only use in a node context, but you're free to explicitly opt in to this request logging. Maybe we could put our `open-api` specific logic in a similar place: https://github.com/globality-corp/leif/pull/6/files#diff-3c5e700f8a6649b789b7421e57b2f74fR79 ?

@jessemyers what do you think about exposing these as part of the logger vs as functions which accept a graph?

It changes the usage in the node project from:
```
app.use(graph.logger.makeMiddleware())
```

To:
```
app.use(makeMiddleware(graph.logger))
```

Do you think one makes for better boundaries over the other? I guess if we keep adding special "make X" functions to the class the logger will get bloated so maybe the latter option is better.

File: dist/index.js (hunk 1 above)
Comment (jessemyers) [reply dist/index.js]: This is correct, it should be git ignored.

It should also probably be called `lib/` for libraries.  This is what our new cookiecutter does.

File: src/logger.js (hunk 6 above)
Comment (jessemyers) [reply src/logger.js]: I don't really "get" what this middleware does with the input logger.

Comment (cdm-ium) [reply src/logger.js]: It generates the per-request audit log:

```
  "length": "1674",
  "message": "graphql",
  "method": "POST",
  "operation-hash": "-",
  "operation-name": "-",
  "request-id": "7363087cdafa4230b150e5545d776af5",
  "request-headers": "{}",
  "response-time": "522.911 ms",
  "status_code": "200",
  "url": "/gql/graphql?",
  "user-id": "-",
  "level": "info"
}
```

**Pull request closed without merging**